subprojects/base-services/src/main/java/org/gradle/internal/reflect/DirectInstantiator.java

```diff
@@ -62,56 +62,85 @@ private DirectInstantiator() {
         if (match == null) {
             throw new IllegalArgumentException(String.format("Could not find any public constructor for %s which accepts parameters %s.", type, Arrays.toString(params)));
         }
-        return match;
+        return match.constructor.get();
     }
 
-    private static boolean isMatch(Constructor<?> constructor, Object... params) {
-        Class<?>[] parameterTypes = constructor.getParameterTypes();
-        if (parameterTypes.length != params.length) {
-            return false;
+    private static class CachedConstructor {
+        private final WeakReference<Constructor<?>> constructor;
+        private final Class<?>[] parameterTypes;
+        private final int paramCount;
+        private final boolean[] isPrimitive;
+        private final Class<?>[] wrappedParameterTypes;
+
+        public CachedConstructor(Constructor<?> ctor) {
+            this.constructor = new WeakReference<Constructor<?>>(ctor);
+            this.parameterTypes = ctor.getParameterTypes();
```

Comment:
It would also be good to collect some infrastructure for this, as we do this kind of inspection in a bunch of places, and they're all subject to various edge cases.